primeqa/mrc/run_mrc.py

```diff
@@ -521,11 +532,12 @@ def process_raw_datasets(raw_datasets, max_samples, training_args, process_fn, s
                 logger.info(f"Filtered EVAL dataset size {eval_examples.num_rows}")
             eval_examples = [eval_examples]
         # Validation feature creation
-        eval_examples, eval_datasets = process_raw_datasets(eval_examples, data_args.max_eval_samples, training_args, preprocessor.process_eval, 'validation')
-        if data_args.eval_fof is not None:
+        eval_examples, eval_datasets = process_raw_datasets(eval_examples, validation_preprocessors, training_args,
+                                                            split='validation', max_samples=data_args.max_eval_samples)
+        if data_args.eval_fof is not None and task_args.task_trainer == MSKD_MRCTrainer:
             eval_dataset = ConcatDataset(eval_datasets)
             setattr(eval_dataset, 'config_name', getattr(eval_dataset.datasets[0], 'config_name'))
-        else:
+        else: # only eval on the first dataset
```

Comment:
Does this mean that for the base trainer, multi-dataset evaluation is handled as evaluation on a single large concatenated dataset?

Reply:
This doesn't match what is said in the README "Multiple datatsets can be included into "--eval_fof", but only the first one is used for evaluation."

Reply:
@ma-sultan I removed the confusing comment. 

@seirasto Using different preprocessor shouldn't crash the training. All our preprocessors for extractive MRC are required to convert their datasets into the same "base" format. So after preprocessing, all featurized datasets can be combined together for training. 

@seirasto @ma-sultan There are two types of multi-dataset MRC training. One is like distillation training, with the trainer 'MSKD_MRCTrainer'. In this case, all datasets are 'homogeneous', which means they use the same preprocessor, postprocessor and eval metric. So all the datasets specified in "eval_fof" can be processed in the same way. The other one is more general, as in this PR, which allows "heterogeneous" datasets to be used. However, HuggingFace trainer class only supports one "post_process_function" and one "compute_metrics". So if we want to evaluate these "heterogeneous" datasets, we have to create different trainers for them. This is doable, but will make run_mrc the code much longer. I would like to create a separate PR for this since this PR is more focused on training side.